[New feature] Migrate old caggs to new form

charislam opened the pull request:
Document procedure, added in TimescaleDB 2.8, that lets you migrate old
caggs (pre-2.7 format) to the new format.

# Links

Fixes #1519 

# Writing help

For information about style and word usage, see the [style guide](https://docs.timescale.com/timescaledb/latest/contribute-to-docs/)

# Review checklists
Reviewers: use this section to ensure you have checked everything before approving this PR:

## Subject matter expert (SME) review checklist

- [ ] Is the content technically accurate?
- [ ] Is the content complete?
- [ ] Is the content presented in a logical order?
- [ ] Does the content use appropriate names for features and products?
- [ ] Does the content provide relevant links to further information?

## Documentation team review checklist

- [x] Is the content free from typos?
- [x] Does the content use plain English?
- [x] Does the content contain clear sections for concepts, tasks, and references?
- [x] Have any images been uploaded to the correct location, and are resolvable?
- [x] If the page index was updated, are redirects required
      and have they been implemented?
- [ ] Have you checked the built version of this content?

charislam committed:
[New feature] Migrate old caggs to new form

Document procedure, added in TimescaleDB 2.8, that lets you migrate old
caggs (pre-2.7 format) to the new format.

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

Loquacity commented on timescaledb/how-to-guides/continuous-aggregates/migrate.md:
```diff
@@ -0,0 +1,19 @@
+with the option `timescaledb.finalized` set to `false`, use the old format.
+
+To migrate a continuous aggregate from the old format to the new format, you can
+use the provided procedure. Connect to your database and run:
```
```suggestion
use this command. Connect to your database and run:
```

It's not really a procedure?

fabriziomello replied:
Yes, it is a procedure. Maybe we can write `use the provided Postgres procedure.` WDYT?

Loquacity replied:
Ah, I see what you mean. A procedure in the Postgres sense, not the documentation sense 😅 

```suggestion
use this procedure. Connect to your database and run:
```

fabriziomello commented on timescaledb/how-to-guides/continuous-aggregates/index.md:
```diff
@@ -20,6 +21,7 @@ only the data that has changed needs to be computed, not the entire dataset.
 *   [Manage materialized hypertables][cagg-mat-hypertables].
 *   [Use real-time aggregates][cagg-realtime].
 *   [Compression with continuous aggregates][cagg-compression].
+*   [Migrate your continuous aggregates][cagg-migrate] from old to new form.
```
```suggestion
*   [Migrate your continuous aggregates][cagg-migrate] from old to new format.
```

fabriziomello commented:
@charislam What about add information about how to check the migration progress/status??? It is possible executing a SELECT query on the `_timescaledb_catalog.continuous_agg_migrate_plan_step`

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

charislam committed:
Add more details about cagg migration

Add details about how to configure the migration and monitor its
progress. Add an API reference entry.

charislam commented on timescaledb/how-to-guides/continuous-aggregates/migrate.md:
```diff
@@ -0,0 +1,43 @@
+with the option `timescaledb.finalized` set to `false`, use the old format.
+
+To migrate a continuous aggregate from the old format to the new format, you can
+use this procedure. It automatically copies over your data and and policies.
```
```suggestion
use this procedure. It automatically copies over your data and policies.
```

charislam committed:
Fix typo

Loquacity commented on timescaledb/how-to-guides/continuous-aggregates/migrate.md:
```diff
@@ -12,8 +12,32 @@ aggregates created in older versions of TimescaleDB, or created in a new version
+
+## Configure continuous aggregate migration
+
+The migration procedure provides two boolean configuration parameters,
```
I think Boolean needs initcaps?

charislam replied:
Not for the abstract data type, according to Google:

> In most contexts, you're talking about a specific data type in a specific programming language, in which case use code font and the exact spelling and capitalization of the programming keyword. But when referring to the abstract data type, use lowercase. (In the unlikely event that you refer to Boolean mathematics or Boolean logic, use uppercase.)

Loquacity replied:
Who knew?! 🤷‍♀️

jfjoly commented on timescaledb/how-to-guides/continuous-aggregates/index.md:
```diff
@@ -20,6 +21,7 @@ only the data that has changed needs to be computed, not the entire dataset.
 *   [Manage materialized hypertables][cagg-mat-hypertables].
 *   [Use real-time aggregates][cagg-realtime].
 *   [Compression with continuous aggregates][cagg-compression].
+*   [Migrate your continuous aggregates][cagg-migrate] from old to new format.
```
Should we be more specific and say `old (pre TimescaleDB 2.7)` ?

charislam replied:
```suggestion
*   [Migrate your continuous aggregates][cagg-migrate] from old to new format.
    Continuous aggregates created in TimescaleDB 2.7 and above are in the new
    format, unless explicitly created in the old format.
```

jfjoly commented on timescaledb/how-to-guides/continuous-aggregates/migrate.md:
```diff
@@ -0,0 +1,43 @@
+
+To migrate a continuous aggregate from the old format to the new format, you can
+use this procedure. It automatically copies over your data and policies.
+
```
We could add here that the procedure is transparent and users can continue to use the continuous aggregate during the migration.

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

charislam commented on timescaledb/how-to-guides/continuous-aggregates/migrate.md:
```diff
@@ -0,0 +1,43 @@
+aggregates created in older versions of TimescaleDB, or created in a new version
+with the option `timescaledb.finalized` set to `false`, use the old format.
+
+To migrate a continuous aggregate from the old format to the new format, you can
+use this procedure. It automatically copies over your data and policies.
```
```suggestion
To migrate a continuous aggregate from the old format to the new format, you can
use this procedure. It automatically copies over your data and policies. You can
continue to use the continuous aggregate while the migration is happening.
```

charislam committed:
Apply suggestions from code review

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

charislam commented:
Direct links for easy previewing:

http://docs-dev.timescale.com/docs-charis-migrate-caggs/timescaledb/charis-migrate-caggs/how-to-guides/continuous-aggregates/migrate/
http://docs-dev.timescale.com/docs-charis-migrate-caggs/api/charis-migrate-caggs/continuous-aggregates/cagg_migrate/

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

jfjoly commented:
Some of the arguments , starting with underscores, are wrong: 
_cagg
_override
_drop_old

Not sure if that gets replaced automatically. The link to the preview pages are not working for me because of a certificate issue. 

This is the UX:
```
-- DEFAULT OPTION: migrate to the new CAGG and rename the old one
CALL cagg_migrate_to_finals_form ('cagg_name');

-- Migrate to the new CAGG and delete the old one
CALL cagg_migrate_to_finals_form ('cagg_name', drop_old_cagg => TRUE);

-- Migrate to the new CAGG with a new name and keep the existing one
CALL cagg_migrate_to_finals_form ('cagg_name', swap_at_end => FALSE);
```

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

Loquacity commented:
> The link to the preview pages are not working for me because of a certificate issue.

Try opening the previews in incognito mode.

jfjoly commented:
Please hold on publishing this as we found an issue. 
https://github.com/timescale/timescaledb/issues/4682

fabriziomello commented:
> Some of the arguments , starting with underscores, are wrong: _cagg _override _drop_old
> 
> Not sure if that gets replaced automatically. The link to the preview pages are not working for me because of a certificate issue.
> 
> This is the UX:
> 
> ```
> -- DEFAULT OPTION: migrate to the new CAGG and rename the old one
> CALL cagg_migrate_to_finals_form ('cagg_name');
> 
> -- Migrate to the new CAGG and delete the old one
> CALL cagg_migrate_to_finals_form ('cagg_name', drop_old_cagg => TRUE);
> 
> -- Migrate to the new CAGG with a new name and keep the existing one
> CALL cagg_migrate_to_finals_form ('cagg_name', swap_at_end => FALSE);
> ```

Actually @jfjoly the parameters starts with underscore

jfjoly commented:
Fabrizio also recommended to add some warning in the docs about the permissions or catalog tables that store the cagg migration state...

I would suggest to add it on the how to guide at the bottom 
For instance:

> **Configure permissions (only required if you get permissions error when executing the migration)**
> The user that execute the procedure should have INSERT/UPDATE grants in the following tables (in the case the user is not a superuser, of course):
> `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan TO %user%;`
> `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan_step TO %user%;`
>

fabriziomello commented:
> Fabrizio also recommended to add some warning in the docs about the permissions or catalog tables that store the cagg migration state...
> 
> I would suggest to add it on the how to guide at the bottom For instance:
> 
> > **Configure permissions (only required if you get permissions error when executing the migration)**
> > The user that execute the procedure should have INSERT/UPDATE grants in the following tables (in the case the user is not a superuser, of course):
> > `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan TO %user%;`
> > `GRANT UPDATE ON _timescaledb_catalog.continuous_agg_migrate_plan_step TO %user%;`

Just a quick fix:

**Configure permissions (only required if you get permissions error when executing the migration)**
The user that execute the procedure should have SELECT/INSERT/UPDATE grants in the following objects (in the case the user is not a superuser, of course):
```sql
GRANT SELECT, INSERT, UPDATE ON TABLE _timescaledb_catalog.continuous_agg_migrate_plan TO %user%;
GRANT SELECT, INSERT, UPDATE ON TABLE _timescaledb_catalog.continuous_agg_migrate_plan_step TO %user%;
GRANT USAGE ON SEQUENCE _timescaledb_catalog.continuous_agg_migrate_plan_step_step_id_seq TO %user%;
```
Where `%user%` should be replaced for the real user name.

charislam committed:
Document troubleshooting

Document troubleshooting for permission errors.

charislam commented:
@fabriziomello @jfjoly Documented! Is it good to merge with this fix?

fabriziomello commented:
> @fabriziomello @jfjoly Documented! Is it good to merge with this fix?

LGTM

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

iroussos commented:
Following the discussion in [slack](https://iobeam.slack.com/archives/C032AS21X4P/p1662822106158919) between @jfjoly and @fabriziomello, there are two different issues to address:

1. The permissions required to perform the migration
2. The parameter names

I agree with @jfjoly on holding any communication, updating the parameter names and then releasing the docs.

@fabriziomello do you mind adding the PR to change the parameter names? 

@charislam We can then also update the parameter names in this PR and include both changes in the next minor release. 

Our only risk is someone using the new undocumented function and hitting the errors. Do you agree with this plan @jfjoly?

fabriziomello commented:
@charislam There are an ongoing PR to fix the `cagg_migrate` parameter names (https://github.com/timescale/timescaledb/pull/4694). I'll let you know once it get merged.

fabriziomello commented:
> @charislam There are an ongoing PR to fix the `cagg_migrate` parameter names ([timescale/timescaledb#4694](https://github.com/timescale/timescaledb/pull/4694)). I'll let you know once it get merged.

Merged. /cc @jfjoly

iroussos commented:
I have a few newbie questions - sorry for the noise but I am still new and trying to understand our processes:

1. @charislam Am I right that docs updates go live the moment we merge them?

    If so, other than updating the parameter names in this PR to follow the fix in https://github.com/timescale/timescaledb/pull/4694, we'll have to wait and merge this PR once the next patch is released. /cc @jfjoly to correct me on this one

2. @charislam Do we somehow mark functions (or functionality) in the docs with the version they are introduced?

    I may missed this, but I don't see any reference to 2.8 in `cagg_migrate`, how are users going to know which version of TimescaleDB they should be running to be able to use the functionality? We mention 2.7 (when the new caggs format was introduced) but not the version the function became available.  
    
    Similarly, how are we going to mark it as 2.8.1+ as the version in 2.8.0 will not work with the params in the documentation?

3. @fabriziomello how's the process of doing a patch release? Is it time consuming, do we try to bundle multiple fixes or do we go wild on patch releases?

    On this specific case, should we start the process for 2.8.1 or is it too early?
    
Thank you!

charislam commented:
> I have a few newbie questions - sorry for the noise but I am still new and trying to understand our processes:
> 
>     1. @charislam Am I right that docs updates go live the moment we merge them?
>        If so, other than updating the parameter names in this PR to follow the fix in [Change parameter names of cagg_migrate procedure timescaledb#4694](https://github.com/timescale/timescaledb/pull/4694), we'll have to wait and merge this PR once the next patch is released. /cc @jfjoly to correct me on this one
> 
>     2. @charislam Do we somehow mark functions (or functionality) in the docs with the version they are introduced?
>        I may missed this, but I don't see any reference to 2.8 in `cagg_migrate`, how are users going to know which version of TimescaleDB they should be running to be able to use the functionality? We mention 2.7 (when the new caggs format was introduced) but not the version the function became available.
>        Similarly, how are we going to mark it as 2.8.1+ as the version in 2.8.0 will not work with the params in the documentation?
> 
>     3. @fabriziomello how's the process of doing a patch release? Is it time consuming, do we try to bundle multiple fixes or do we go wild on patch releases?
>        On this specific case, should we start the process for 2.8.1 or is it too early?
> 
> 
> Thank you!

@iroussos Those are very, very good questions!

1. Yes, they go live immediately, so if the patch is not out yet, we should wait for it to go out.
2. We do not have versioning currently in the docs. Adding 'introduced in version' labels is on the backlog, for exactly this reason! The first rendition will probably just be a label on the entire API reference. Figuring out what to do with renamed parameters is a more complex question and we'll have to do some thinking about how we display that info for readers given we don't have full-scale versioning

iroussos commented on api/cagg_migrate.md:
```diff
@@ -0,0 +1,51 @@
+
+|Name|Type|Description|
+|-|-|-|
+|`_cagg`|`REGCLASS`|The continuous aggregate to migrate|
```
```suggestion
|`cagg`|`REGCLASS`|The continuous aggregate to migrate|
```

iroussos commented on api/cagg_migrate.md:
```diff
@@ -0,0 +1,51 @@
+
+|Name|Type|Description|
+|-|-|-|
+|`_override`|`BOOLEAN`|If false, the old continuous aggregate keeps its name. The new continuous aggregate is named `<OLD_CONTINUOUS_AGGREGATE_NAME>_new`. If true, the new continuous aggregate gets the old name. The old continuous aggregate is renamed `<OLD_CONTINUOUS_AGGREGATE_NAME>_old`. Defaults to `false`.|
```
```suggestion
|`override`|`BOOLEAN`|If false, the old continuous aggregate keeps its name. The new continuous aggregate is named `<OLD_CONTINUOUS_AGGREGATE_NAME>_new`. If true, the new continuous aggregate gets the old name. The old continuous aggregate is renamed `<OLD_CONTINUOUS_AGGREGATE_NAME>_old`. Defaults to `false`.|
```

iroussos commented on api/cagg_migrate.md:
```diff
@@ -0,0 +1,51 @@
+|Name|Type|Description|
+|-|-|-|
+|`_override`|`BOOLEAN`|If false, the old continuous aggregate keeps its name. The new continuous aggregate is named `<OLD_CONTINUOUS_AGGREGATE_NAME>_new`. If true, the new continuous aggregate gets the old name. The old continuous aggregate is renamed `<OLD_CONTINUOUS_AGGREGATE_NAME>_old`. Defaults to `false`.|
+|`_drop_old`|`BOOLEAN`|If true, the old continuous aggregate is deleted. Must be used together with `_override`. Defaults to `false`.|
```
```suggestion
|`drop_old`|`BOOLEAN`|If true, the old continuous aggregate is deleted. Must be used together with `override`. Defaults to `false`.|
```

iroussos commented on timescaledb/how-to-guides/continuous-aggregates/migrate.md:
```diff
@@ -0,0 +1,52 @@
+## Configure continuous aggregate migration
+
+The migration procedure provides two boolean configuration parameters,
+`_override` and `_drop_old`. By default, the name of your new continuous
+aggregate is the name of your old continuous aggregate, with the suffix `_new`.
+
+Set `_override` to true to rename your new continuous aggregate with the
+original name. The old continuous aggregate is renamed with the suffix `_old`.
+
+To both rename and drop the old continuous aggregate entirely, set both
+parameters to true. Note that `_drop_old` must be used together with
+`_override`.
```
```suggestion
`override` and `drop_old`. By default, the name of your new continuous
aggregate is the name of your old continuous aggregate, with the suffix `_new`.

Set `override` to true to rename your new continuous aggregate with the
original name. The old continuous aggregate is renamed with the suffix `_old`.

To both rename and drop the old continuous aggregate entirely, set both
parameters to true. Note that `drop_old` must be used together with
`override`.
```

iroussos reviewed (changes requested):
@charislam I have a few proposed changes to update the parameter names in this PR to follow the fix in https://github.com/timescale/timescaledb/pull/4694

Do you mind checking if everything is right and updating the PR?

We plan to move forward with the release of 2.8.1 (https://github.com/timescale/eng-database/issues/321), so it would be great to have the docs PR ready to be merged

charislam committed:
Rename parameters

Co-authored-by: Yannis Roussos <iroussos@gmail.com>

charislam committed:
Merge branch 'latest' into charis/migrate-caggs

charislam commented:
> @charislam I have a few proposed changes to update the parameter names in this PR to follow the fix in [timescale/timescaledb#4694](https://github.com/timescale/timescaledb/pull/4694)
> 
> Do you mind checking if everything is right and updating the PR?
> 
> We plan to move forward with the release of 2.8.1 ([timescale/eng-database#321](https://github.com/timescale/eng-database/issues/321)), so it would be great to have the docs PR ready to be merged

@iroussos everything looks good! I committed your suggested changes and solved the merge conflicts, so we should be ready to merge whenever you give the 👍🏼

iroussos commented:
Thank you @charislam, much appreciated 🚀

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

charislam commented:
@iroussos I saw that 2.8.1 is tagged, am I good to merge this now?

iroussos commented:
@charislam I agree with @fabriziomello, we'll ping you once 2.8.1 is publicly available to our cloud users

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs

iroussos reviewed (approved):
@charislam on our side we are ready to merge the doc changes; 2.8.1 has been released to MST and will be released as well to Timescale cloud today.

/cc @jnidzwetzki as the release captain

Loquacity committed:
Merge branch 'latest' into charis/migrate-caggs
